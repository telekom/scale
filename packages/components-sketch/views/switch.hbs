{{!--
Scale https://github.com/telekom/scale

Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
--}}

<div data-sketch-artboard="Switch" class="artboard">
  <h1 class="title--artboard">Switch</h1>

  <h2 class="title--artboard-section">Selected On</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-switch
        label="Switch Label"
        checked
        data-sketch-key="switch-1"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-switch
        label="Switch Label"
        checked
        data-sketch-state="hover"
        data-sketch-key="switch-2"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-switch
        label="Switch Label"
        checked
        data-sketch-state="focus"
        data-sketch-key="switch-3"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-switch
        label="Switch Label"
        checked
        data-sketch-state="active"
        data-sketch-key="switch-4"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
      <scale-switch
        label="Switch Label"
        checked
        disabled
        data-sketch-key="switch-5"
      ></scale-switch>
    </div>
  </div>

  <h2 class="title--artboard-section">Selected Off</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-switch
        label="Switch Label"
        data-sketch-key="switch-10"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-switch
        label="Switch Label"
        data-sketch-state="hover"
        data-sketch-key="switch-6"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-switch
        label="Switch Label"
        data-sketch-state="focus"
        data-sketch-key="switch-7"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-switch
        label="Switch Label"
        data-sketch-state="active"
        data-sketch-key="switch-8"
      ></scale-switch>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
      <scale-switch
        label="Switch Label"
        disabled
        data-sketch-key="switch-9"
      ></scale-switch>
    </div>
  </div>
</div>

<script src="/symbol_names.js"></script>
<script>
  Array.from(document.querySelectorAll('scale-switch')).forEach((btn) => {
    btn.dataset.sketchSymbol = `Switch / ${order(
      1,
      btn.hasAttribute('checked') ? 'Selected On' : 'Selected Off'
    )} / ${order(4, getStateName(btn))}`;
  });
</script>
