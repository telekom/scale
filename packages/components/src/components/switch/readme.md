# t-switch



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                | Type      | Default |
| ------------- | -------------- | -------------------------- | --------- | ------- |
| `active`      | `active`       | (optional) Active switch   | `boolean` | `false` |
| `customClass` | `custom-class` | (optional) Switch class    | `string`  | `''`    |
| `disabled`    | `disabled`     | (optional) Disabled switch | `boolean` | `false` |
| `theme`       | `theme`        | (optional) Switch theme    | `string`  | `''`    |


## CSS Custom Properties

| Name                                | Description                            |
| ----------------------------------- | -------------------------------------- |
| `--switch-active-after-left`        | Left of the switch active after        |
| `--switch-active-after-margin-left` | Margin left of the switch active after |
| `--switch-active-background-color`  | Background color of the switch active  |
| `--switch-active-border-color`      | Border color of the switch active      |
| `--switch-after-background-color`   | Background color of the switch after   |
| `--switch-after-border-radius`      | Border radius of the switch after      |
| `--switch-after-content`            | Content of the switch after            |
| `--switch-after-height`             | Height of the switch after             |
| `--switch-after-left`               | Left of the switch after               |
| `--switch-after-position`           | Position of the switch after           |
| `--switch-after-top`                | Top of the switch after                |
| `--switch-after-transition`         | Transition radius of the switch after  |
| `--switch-after-width`              | Width of the switch after              |
| `--switch-background`               | Background of the switch               |
| `--switch-border`                   | Border of the switch                   |
| `--switch-border-radius`            | Border radius of the switch            |
| `--switch-cursor`                   | Cursor of the switch                   |
| `--switch-disabled-cursor`          | Cursor of the switch disabled          |
| `--switch-disabled-opacity`         | Opacity of the switch disabled         |
| `--switch-display`                  | Display of the switch                  |
| `--switch-height`                   | Height of the switch                   |
| `--switch-margin`                   | Margin of the switch                   |
| `--switch-outline`                  | Outline of the switch                  |
| `--switch-position`                 | Position of the switch                 |
| `--switch-transition`               | Transition radius of the switch        |
| `--switch-vertical-align`           | Vertical Align of the switch           |
| `--switch-width`                    | Width of the switch                    |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
