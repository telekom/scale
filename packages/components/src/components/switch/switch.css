/**
* @prop --switch-margin: Margin of the switch
* @prop --switch-display: Display of the switch
* @prop --switch-position: Position of the switch
* @prop --switch-width: Width of the switch
* @prop --switch-height: Height of the switch
* @prop --switch-border: Border of the switch
* @prop --switch-outline: Outline of the switch
* @prop --switch-border-radius: Border radius of the switch
* @prop --switch-background: Background of the switch
* @prop --switch-cursor: Cursor of the switch
* @prop --switch-transition: Transition radius of the switch
* @prop --switch-vertical-align: Vertical Align of the switch
* @prop --switch-after-content: Content of the switch after
* @prop --switch-after-position: Position of the switch after
* @prop --switch-after-top: Top of the switch after
* @prop --switch-after-left: Left of the switch after
* @prop --switch-after-border-radius: Border radius of the switch after
* @prop --switch-after-transition: Transition radius of the switch after
* @prop --switch-after-width: Width of the switch after
* @prop --switch-after-height: Height of the switch after
* @prop --switch-after-background-color: Background color of the switch after
* @prop --switch-active-border-color: Border color of the switch active
* @prop --switch-active-background-color: Background color of the switch active
* @prop --switch-active-after-left: Left of the switch active after
* @prop --switch-active-after-margin-left: Margin left of the switch active after
* @prop --switch-disabled-cursor: Cursor of the switch disabled
* @prop --switch-disabled-opacity: Opacity of the switch disabled
*/

.switch {
  margin: var(--switch-margin, 0);
  display: var(--switch-display, inline-block);
  position: var(--switch-position, relative);
  width: var(--switch-width, 40px);
  height: var(--switch-height, 20px);
  border: var(--switch-border, 1px solid #dcdfe6);
  outline: var(--switch-outline, none);
  border-radius: var(--switch-border-radius, 10px);
  background: var(--switch-background, #dcdfe6);
  cursor: var(--switch-cursor, pointer);
  transition: var(
    --switch-transition,
    border-color 0.3s,
    background-color 0.3s
  );
  vertical-align: var(--switch-vertical-align, middle);
}

.switch:after {
  content: var(--switch-after-content, '');
  position: var(--switch-after-position, absolute);
  top: var(--switch-after-top, 2px);
  left: var(--switch-after-left, 1px);
  border-radius: var(--switch-after-border-radius, 100%);
  transition: var(--switch-after-transition, all 0.3s);
  width: var(--switch-after-width, 16px);
  height: var(--switch-after-height, 16px);
  background-color: var(--switch-after-background-color, #fff);
}

.switch--active {
  border-color: var(--switch-active-border-color, #409eff);
  background-color: var(--switch-active-background-color, #409eff);
}

.switch--active:after {
  left: var(--switch-active-after-left, 100%);
  margin-left: var(--switch-active-after-margin-left, -17px);
}

.switch--disabled {
  cursor: var(--switch-disabled-cursor, not-allowed);
  opacity: var(--switch-disabled-opacity, 0.5);
}
