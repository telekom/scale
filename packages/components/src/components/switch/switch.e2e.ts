import { newE2EPage } from '@stencil/core/testing';

describe('t-switch', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-switch></t-switch>');
    const element = await page.find('t-switch');
    expect(element).toHaveClass('hydrated');
  });

  it('should not toggle to active when switch is disabled ', async () => {
    const page = await newE2EPage({
      html: `<t-switch disabled="true" active="false"></t-switch>`,
    });
    const ele = await page.find('t-switch >>> .switch');
    await ele.click();
    expect(ele).not.toContain('.switch--active');
  });
});
