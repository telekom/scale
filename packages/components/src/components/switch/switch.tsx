import { Component, h, Prop } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';

@Component({
  tag: 't-switch',
  styleUrl: 'switch.css',
  shadow: true,
})
export class Switch {
  /** (optional) Switch class */
  @Prop() public customClass?: string = '';
  /** (optional) Switch theme */
  @Prop() public theme?: string = '';
  /** (optional) Active switch */
  @Prop() public active?: boolean = false;
  /** (optional) Disabled switch */
  @Prop() public disabled?: boolean = false;

  public toggleSwitch = () => {
    if (this.disabled) {
      return;
    }
    this.active = !this.active;
  };

  public render() {
    return <div class={this.getCssClassMap()} onClick={this.toggleSwitch} />;
  }

  private getCssClassMap(): CssClassMap {
    return classNames(
      'switch',
      this.customClass && this.customClass,
      this.theme && `alert--theme-${this.theme}`,
      this.active && `switch--active`,
      this.disabled && `switch--disabled`
    );
  }
}
