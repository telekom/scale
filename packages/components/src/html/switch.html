<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <title>Stencil Component Starter</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
    <style type="text/css" media="screen, print">
      html {
        margin: 0;
        padding: 0;
      }

      body {
        margin: 0;
        padding: 4rem;
      }
    </style>
  </head>

  <body>
    <p>
      <scale-switch
        id="toggle-disabled"
        label="Disable all below"
      ></scale-switch>
    </p>

    <p>
      <scale-switch label="Hello friend (web)"></scale-switch>
    </p>

    <p data-platform="ios">
      <scale-switch label="Hello friend (ios)"></scale-switch>
    </p>

    <p data-platform="android">
      <scale-switch label="Hello friend (android)"></scale-switch>
    </p>

    <script>
      let t = document.querySelector('#toggle-disabled');
      let all = [...document.querySelectorAll('scale-switch')].filter(
        (x) => x !== t
      );
      t.addEventListener('scale-change', (event) => {
        all.forEach((x) => {
          x.disabled = event.target.checked;
        });
      });
    </script>
  </body>
</html>
