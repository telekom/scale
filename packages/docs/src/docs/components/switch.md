---
title: Switch
description: Switch element
url: /docs/switch
contributors:
  - SylviaGonschior
---

# Switch

Switch component

selector: `t-switch`

## Default

```javascript
<t-switch></t-switch>
```

<div class="demo-container">
  <t-switch></t-switch>
</div>

## Switch disabled

```javascript
<t-switch active="true" disabled="true"></t-switch>
```

<div class="demo-container">
   <t-switch active="true" disabled="true"></t-switch>
</div>

```javascript
<t-switch active="false" disabled="true"></t-switch>
```

<div class="demo-container">
   <t-switch active="false" disabled="true"></t-switch>
</div>
