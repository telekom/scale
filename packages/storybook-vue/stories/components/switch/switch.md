<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Switch</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>

Switches allow users to quickly turn an item on or off. Like a physical light switch, they provide an immediate result.

## General

A switch offers two mutually exclusive states (on/off, yes/no). It always has a default value. A solid Magenta color shows it is on, while a dark outline shows it is off.

![Image Name](assets/3_components/switch/switch.png)

### When to use

Use a switch whenever the user may want to immediately activate or deactivate something. It is commonly used as an on/off switch, e.g., to control preferences.

### When not to use

Use radio buttons or a single checkbox if

- the user has to confirm by saving or submitting to apply the new state or
- immediate results are not possible.

## Label

Clearly label the switch and keep it short. For some use cases, a single noun provides the best guidance.

![Image Name](assets/3_components/switch/switch_label.png)

## Related components

<a href="?path=/usage/components-checkbox--standard">Checkbox, </a>
<a href="?path=/usage/components-dropdown--standard">Dropdown, </a>
<a href="?path=/usage/components-radio-button--standard">Radio Button</a>
