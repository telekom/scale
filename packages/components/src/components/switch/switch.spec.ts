import { newSpecPage } from '@stencil/core/testing';
import { Switch } from './switch';

describe('Switch', () => {
  let element;
  beforeEach(async () => {
    element = new Switch();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Switch],
      html: `<t-switch>/t-switch>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should have class active', () => {
    element.active = true;
    expect(element.getCssClassMap()).toContain('switch--active');
  });

  it('should have class disabled', () => {
    element.disabled = true;
    expect(element.getCssClassMap()).toContain('switch--disabled');
  });

  it('should have toggle active state', () => {
    element.disabled = false;
    element.active = true;
    element.toggleSwitch();
    expect(element.active).toBe(false);
  });

  it('should not toggle active state if disabled', () => {
    element.disabled = true;
    element.active = true;
    element.toggleSwitch();
    expect(element.active).toBe(true);
  });
});
