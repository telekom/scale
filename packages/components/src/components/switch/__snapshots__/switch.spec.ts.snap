// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should match snapshot 1`] = `
<t-switch>
  <mock:shadow-root>
    <div class="switch"></div>
  </mock:shadow-root>
  /t-switch&gt;
</t-switch>
`;
